### PR TITLE
Shows assemblies some love. ♥

### LINKED ISSUE #27235: Attached proximity sensors are broken
The timer to arm proximity sensors does not count down, and even when manually armed, proximity sensors do not trigger when people walk by.

### LINKED ISSUE #35829: signaller +IR beam and signaller + proxy sensor assemblies don't work
Issue reported from  Round ID: 83879 (/tg/Station Bagil [ENGLISH] [US-EAST] [100% FREE LAG])

Reporting client version: 512

In addition, proxy sensors dont work when attached to a wire

Strangely, voice analyzer + signaller assemblies work just fine

### LINKED ISSUE #37545: Proximity sensors don't trigger when attached to wires
Revision: de523243b8211bb57c7a89e0f7e312fa6b228cda
/obj/item/assembly/prox_sensor

If you attach this to a wire (for example an airlock's), it won't trigger once someone enters the detection range.

### PR DESCRIPTION
**Pull request by ShizCalev**
🆑 ShizCalev
fix: Fixed proximity sensors not working
fix: Fixed proximity sensors showing the arming timer when it's already armed
fix: Fixed mousetraps not working properly when attached to TTVs and one tank bombs.
fix: Fixed mousetraps losing their inhands when armed
fix: Fixed infrared sensors not updating the direction of their beam when connected to another device.
fix: Fixed infrared sensor beams not aligning correctly with the device when attached to another device, or a TTV
fix: Fixed infrared sensors not turning on when connected to another device....
fix: Fixed infrared sensors facing the wrong direction when connected to another assembly device.
fix: Fixed infrared sensor beams not starting at the actually emitter portion of the sensor.
fix: Fixed infrared sensor beams being a pixel shorter than the emitter.
fix: Fixed infrared sensor beams having a two off-colored pixels for most directions.
fix: Fixed infrared sensors completely breaking when part of a single tank bomb.
fix: Fixed assembly items being detached from one another dropping both items at the incorrect location.
fix: Fixed deaf mobs being able to hear pAI signalers and health sensors.
fix: Fixed single tank bombs not correctly updating their overlays when turning an object on in the assembly.
fix: Fixed single tank bombs not flashing the user when you weld them.
tweak: You can now actually HEAR the beeping when a sensor or signaler is triggered!
/🆑

Fixes #35829
Fixes #27235
Fixes #37545

Definitely room for improvement still, but this gets things working at the least.

### CODE REVIEW

**Comment by optimumtact** [comment]
Great PR :heart:

**Comment by MrDoomBringer** [comment]
holy shit
i can finally code a real proximity suicide now

**Comment by PKPenguin321** [comment]
Might be out of scope but this just reminded me to report https://github.com/tgstation/tgstation/issues/37635 and a fix for that would be appropriate here

**Comment by ShizCalev** [comment]
@PKPenguin321 think the fix for that would be the componentization of flashes.

**Comment by Iamgoofball** [comment]
Assemblies need to be axed entirely.

**Comment by SpaceManiac** [comment]
>fix: Fixed proximity sensors not working

THANK you

**Comment by AlexTheSysop42** [comment]
Bug: signallers have issues when put in buttons

File: code/game/objects/items/devices/transfer_valve.dm [hunk 1]
```diff
@@ -57,6 +57,35 @@
...
+	. = ..()
+	if(attached_device)
+		attached_device.dropped()
+
+/obj/item/transfer_valve/on_found(mob/finder)
+	if(attached_device)
+		attached_device.on_found(finder)
+
+/obj/item/transfer_valve/Crossed(atom/movable/AM as mob|obj)
+	if(attached_device)
```
Comment (Cyberboss) [inline code/game/objects/items/devices/transfer_valve.dm]: . = ..()

File: code/modules/assembly/doorcontrol.dm [hunk 2]
```diff
@@ -2,26 +2,26 @@
...
+	cooldown = TRUE
 	var/openclose
 	for(var/obj/machinery/door/poddoor/M in GLOB.machines)
 		if(M.id == src.id)
 			if(openclose == null)
 				openclose = M.density
 			INVOKE_ASYNC(M, openclose ? /obj/machinery/door/poddoor.proc/open : /obj/machinery/door/poddoor.proc/close)
 	sleep(10)
-	cooldown = 0
+	cooldown = FALSE
```
Comment (Cyberboss) [inline code/modules/assembly/doorcontrol.dm]: use a timer with VARSET_CALLBACK please

File: code/modules/assembly/doorcontrol.dm [hunk 3]
```diff
@@ -67,15 +67,15 @@
 		INVOKE_ASYNC(D, doors_need_closing ? /obj/machinery/door/airlock.proc/close : /obj/machinery/door/airlock.proc/open)
 
 	sleep(10)
-	cooldown = 0
+	cooldown = FALSE
```
Comment (Cyberboss) [inline code/modules/assembly/doorcontrol.dm]: ^

File: code/modules/assembly/doorcontrol.dm [hunk 4]
```diff
@@ -93,15 +93,15 @@
 			INVOKE_ASYNC(M, /obj/machinery/door/poddoor.proc/close)
 
 	sleep(10)
-	cooldown = 0
+	cooldown = FALSE
```
Comment (Cyberboss) [inline code/modules/assembly/doorcontrol.dm]: ^

File: code/modules/assembly/doorcontrol.dm [hunk 5]
```diff
@@ -113,32 +113,32 @@
 			M.icon_state = "igniter[M.on]"
 
 	sleep(30)
-	cooldown = 0
+	cooldown = FALSE
```
Comment (Cyberboss) [inline code/modules/assembly/doorcontrol.dm]: ^

File: code/modules/assembly/doorcontrol.dm [hunk 6]
```diff
@@ -113,32 +113,32 @@
...
 /obj/item/assembly/control/flasher/activate()
-	cooldown = 1
+	cooldown = TRUE
 	for(var/obj/machinery/flasher/M in GLOB.machines)
 		if(M.id == src.id)
 			INVOKE_ASYNC(M, /obj/machinery/flasher.proc/flash)
 
 	sleep(50)
-	cooldown = 0
+	cooldown = FALSE
```
Comment (Cyberboss) [inline code/modules/assembly/doorcontrol.dm]: ^

File: code/modules/assembly/doorcontrol.dm [hunk 7]
```diff
@@ -113,32 +113,32 @@
...
 /obj/item/assembly/control/crematorium/activate()
-	cooldown = 1
+	cooldown = TRUE
 	for (var/obj/structure/bodycontainer/crematorium/C in GLOB.crematoriums)
 		if (C.id == id)
 			C.cremate(usr)
 
 	sleep(50)
-	cooldown = 0
+	cooldown = FALSE
```
Comment (Cyberboss) [inline code/modules/assembly/doorcontrol.dm]: ^

File: code/modules/assembly/infrared.dm [hunk 8]
```diff
@@ -103,19 +135,15 @@
...
 /obj/item/assembly/infra/proc/trigger_beam(atom/movable/AM, turf/location)
 	refreshBeam()
 	switchListener(location)
 	if(!secured || !on || next_activate > world.time)
 		return FALSE
-	pulse(0)
-	audible_message("[icon2html(src, hearers(src))] *beep* *beep*", null, 3)
+	pulse(FALSE)
+	audible_message("[icon2html(src, hearers(src))] *beep* *beep* *beep*", null, hearing_range)
+	for(var/mob/CHM in get_hearers_in_view(hearing_range, src))
```
Comment (Cyberboss) [inline code/modules/assembly/infrared.dm]: typeless

File: code/modules/assembly/proximity.dm [hunk 9]
```diff
@@ -51,25 +55,29 @@
...
-/obj/item/assembly/prox_sensor/sense()
+/obj/item/assembly/prox_sensor/proc/sense()
 	if(!scanning || !secured || next_activate > world.time)
-		return 0
-	pulse(0)
-	audible_message("[icon2html(src, hearers(src))] *beep* *beep*", null, 3)
+		return FALSE
+	pulse(FALSE)
+	audible_message("[icon2html(src, hearers(src))] *beep* *beep* *beep*", null, hearing_range)
+	for(var/mob/CHM in get_hearers_in_view(hearing_range, src))
```
Comment (Cyberboss) [inline code/modules/assembly/proximity.dm]: typeless

File: code/modules/assembly/timer.dm [hunk 10]
```diff
@@ -53,20 +56,23 @@
 /obj/item/assembly/timer/proc/timer_end()
 	if(!secured || next_activate > world.time)
 		return FALSE
-	pulse(0)
-	audible_message("[icon2html(src, hearers(src))] *beep* *beep*", null, 3)
+	pulse(FALSE)
+	audible_message("[icon2html(src, hearers(src))] *beep* *beep* *beep*", null, hearing_range)
+	for(var/mob/CHM in get_hearers_in_view(hearing_range, src))
```
Comment (Cyberboss) [inline code/modules/assembly/timer.dm]: typelass

File: code/modules/integrated_electronics/subtypes/input.dm [hunk 11]
```diff
@@ -690,7 +691,9 @@
 		return 0
 
 	activate_pin(3)
-	audible_message("[icon2html(src, hearers(src))] *beep* *beep*", null, 1)
+	audible_message("[icon2html(src, hearers(src))] *beep* *beep* *beep*", null, hearing_range)
+	for(var/mob/CHM in get_hearers_in_view(hearing_range, src))
```
Comment (Cyberboss) [inline code/modules/integrated_electronics/subtypes/input.dm]: .

**Review by Cyberboss — CHANGES_REQUESTED** [review]

**Pull request merged**